libs/contrib/Control/Algebra/Laws.idr

```diff
@@ -0,0 +1,332 @@
+module Control.Algebra.Laws
+
+import Prelude
+import Control.Algebra
+
+%default total
+
+-- Monoids
+
+||| Inverses are unique.
+public export
+uniqueInverse : VMonoid ty => (x, y, z : ty) ->
+  y <+> x = neutral {ty} -> x <+> z = neutral {ty} -> y = z
+uniqueInverse x y z p q =
+  rewrite sym $ monoidNeutralIsNeutralL y in
+    rewrite sym q in
+      rewrite semigroupOpIsAssociative y x z in
+  rewrite p in
+    rewrite monoidNeutralIsNeutralR z in
+      Refl
+
+-- Groups
+
+||| Only identity is self-squaring.
+public export
+selfSquareId : VGroup ty => (x : ty) ->
+  x <+> x = x -> x = neutral {ty}
+-- selfSquareId x p =
```

Comment:
Using Idris2 from 4aa1c1f44, this definition, when uncommented, compiles for me.

Reply:
Using 6898965a:
```
Control/Algebra/Laws.idr:33:7--35:1:While processing right hand side of selfSquareId at Control/Algebra/Laws.idr:28:1--35:1:
Can't solve constraint between:
	neutral
and
	inverse x <+> x
```